azurerm/resource_arm_key_vault_secret.go

```diff
@@ -0,0 +1,240 @@
+	tags := d.Get("tags").(map[string]interface{})
+
+	parameters := keyvault.SecretSetParameters{
+		Value:       azure.String(value),
+		ContentType: azure.String(contentType),
+		Tags:        expandTags(tags),
+	}
+
+	_, err := client.SetSecret(keyVaultBaseUrl, name, parameters)
+	if err != nil {
+		return err
+	}
+
+	// "" indicates the latest version
+	read, err := client.GetSecret(keyVaultBaseUrl, name, "")
+	if err != nil {
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read KeyVault Secret '%s' (in key vault '%s')", name, keyVaultBaseUrl)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmKeyVaultSecretRead(d, meta)
+}
+
+func resourceArmKeyVaultSecretUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).keyVaultManagementClient
+	log.Print("[INFO] preparing arguments for AzureRM KeyVault Secret update.")
+
+	id, err := parseKeyVaultSecretID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	value := d.Get("value").(string)
+	contentType := d.Get("content_type").(string)
+	tags := d.Get("tags").(map[string]interface{})
+
+	if d.HasChange("value") {
+		// for changing the value of the secret we need to create a new version
+		parameters := keyvault.SecretSetParameters{
+			Value:       azure.String(value),
+			ContentType: azure.String(contentType),
+			Tags:        expandTags(tags),
+		}
+
+		_, err := client.SetSecret(id.KeyVaultBaseUrl, id.Name, parameters)
+		if err != nil {
+			return err
+		}
+
+		// "" indicates the latest version
+		read, err := client.GetSecret(id.KeyVaultBaseUrl, id.Name, "")
+		id, err = parseKeyVaultSecretID(*read.ID)
+		if err != nil {
+			return err
+		}
+
+		// the ID is suffixed with the secret version
+		d.SetId(*read.ID)
+	} else {
+		parameters := keyvault.SecretUpdateParameters{
+			ContentType: azure.String(contentType),
+			Tags:        expandTags(tags),
+		}
+
+		_, err = client.UpdateSecret(id.KeyVaultBaseUrl, id.Name, id.Version, parameters)
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceArmKeyVaultSecretRead(d, meta)
+}
+
+func resourceArmKeyVaultSecretRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).keyVaultManagementClient
+
+	id, err := parseKeyVaultSecretID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	// we always want to get the latest version
+	resp, err := client.GetSecret(id.KeyVaultBaseUrl, id.Name, "")
+	if err != nil {
+		if responseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on Azure KeyVault Secret %s: %+v", id.Name, err)
+	}
+
+	// the version may have changed, so parse the updated id
+	respID, err := parseKeyVaultSecretID(*resp.ID)
+	if err != nil {
+		return err
+	}
+
+	d.Set("name", respID.Name)
+	d.Set("vault_uri", respID.KeyVaultBaseUrl)
+	d.Set("value", resp.Value)
+	d.Set("version", respID.Version)
+	d.Set("content_type", resp.ContentType)
+
+	flattenAndSetTags(d, resp.Tags)
```

Comment:
complex types should always check the returned error from `d.Set()`, more opportunities for things to go wrong

Reply:
given this affects every resource in the Provider - I'm going to make this a separate task & do this across everything